### **NLP Final Project**
#### **Spam and Sentiment Email Analysis: Spam Bi-LSTM Supervised Learning**

Wilson Neira

##### **1. Import**
* Import libraries needed for deep learning and text sequence preparation with TensorFlow/Keras.

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.metrics import classification_report
from sklearn.model_selection import StratifiedKFold
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, Embedding, LSTM, Bidirectional, Dense, Concatenate


##### **2. Tokenization and Sequence Padding**

* Load datasets, encode labels, convert email texts into padded numeric sequences using Tokenizer.

In [2]:
# Load Data
train_df = pd.read_csv("group34_train_data_with_clusters.csv")
test_df = pd.read_csv("group34_test_data_with_clusters.csv")

# Encode labels
le = LabelEncoder()
train_labels = le.fit_transform(train_df['label'])  # spam:1, ham:0
test_labels = le.transform(test_df['label'])

# Prepare tokenizer (fit on train)
tokenizer = Tokenizer(num_words=5000)
tokenizer.fit_on_texts(train_df['email'])

# Text to sequences
X_train_seq = tokenizer.texts_to_sequences(train_df['email'])
X_test_seq = tokenizer.texts_to_sequences(test_df['email'])

# Padding sequences
max_len = 200
X_train_pad = pad_sequences(X_train_seq, maxlen=max_len, padding='post')
X_test_pad = pad_sequences(X_test_seq, maxlen=max_len, padding='post')


##### **3. Bi-LSTM Model Definition**
* Define a flexible Bi-LSTM model that optionally incorporates clustering features through concatenation for classification.

In [3]:
# Function to create Bi-LSTM model
def create_model(cluster_feature_dim=0):
    input_text = Input(shape=(max_len,))
    embedding = Embedding(input_dim=5000, output_dim=128)(input_text)
    x = Bidirectional(LSTM(64))(embedding)

    if cluster_feature_dim > 0:
        input_cluster = Input(shape=(cluster_feature_dim,))
        concatenated = Concatenate()([x, input_cluster])
        output = Dense(1, activation='sigmoid')(concatenated)
        model = Model(inputs=[input_text, input_cluster], outputs=output)
    else:
        output = Dense(1, activation='sigmoid')(x)
        model = Model(inputs=input_text, outputs=output)

    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

##### **4. Prepare Clustering Features and Configurations**
* Perform one-hot encoding on K-Means and hierarchical cluster labels, prepare datasets for each model configuration and set up the StratifiedKFold for cross-validation.

In [ ]:
# Prepare clustering features
cluster_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
train_kmeans_feat = cluster_encoder.fit_transform(train_df[['kmeans_cluster']])
test_kmeans_feat = cluster_encoder.transform(test_df[['kmeans_cluster']])

train_hier_feat = cluster_encoder.fit_transform(train_df[['hierarchical_cluster']])
test_hier_feat = cluster_encoder.transform(test_df[['hierarchical_cluster']])

# K-fold Cross-validation
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Store configurations with corresponding features
configurations = {
    'Baseline Bi-LSTM': (X_train_pad, None, None),
    'Bi-LSTM + K-Means': (X_train_pad, train_kmeans_feat, test_kmeans_feat),
    'Bi-LSTM + Hierarchical': (X_train_pad, train_hier_feat, test_hier_feat)
}

##### **5. Stratified 5-Fold Cross-Validation**
* Implement a 5-fold cross-validation strategy for each model configuration, train and validate the models and report classification metrics for each configuration.

In [ ]:
# Perform StratifiedKFold cross-validation for each configuration
for config_name, (text_data, cluster_data, test_cluster_data) in configurations.items():
    print(f"\n--- Stratified 5-Fold CV: {config_name} ---")

    all_val_preds = []
    all_val_actuals = []
    fold = 1
    for train_index, val_index in skf.split(text_data, train_labels):
        print(f"\nFold {fold}")
        X_text_train, X_text_val = text_data[train_index], text_data[val_index]
        y_train_fold, y_val_fold = train_labels[train_index], train_labels[val_index]

        # Prepare clustering data if possible
        if cluster_data is not None:
            X_cluster_train, X_cluster_val = cluster_data[train_index], cluster_data[val_index]
            model = create_model(cluster_feature_dim=cluster_data.shape[1])
            model.fit(
                [X_text_train, X_cluster_train], y_train_fold,
                epochs=3, batch_size=64,
                validation_data=([X_text_val, X_cluster_val], y_val_fold)
            )
            val_pred = (model.predict([X_text_val, X_cluster_val]) > 0.5).astype("int32")
        else:
            model = create_model()
            model.fit(
                X_text_train, y_train_fold,
                epochs=3, batch_size=64,
                validation_data=(X_text_val, y_val_fold)
            )
            val_pred = (model.predict(X_text_val) > 0.5).astype("int32")

        all_val_preds.extend(val_pred.flatten())
        all_val_actuals.extend(y_val_fold)
        fold += 1

    # Classification report after cross-validation for each configuration
    print(f"\nClassification Report for {config_name} (5-fold CV combined results):")
    print(classification_report(all_val_actuals, all_val_preds, target_names=le.classes_))


--- Stratified 5-Fold CV: Baseline Bi-LSTM ---

Fold 1
Epoch 1/3
338/338 ━━━━━━━━━━━━━━━━━━━━ 12s 22ms/step - accuracy: 0.8769 - loss: 0.2958 - val_accuracy: 0.9822 - val_loss: 0.0562
Epoch 2/3
338/338 ━━━━━━━━━━━━━━━━━━━━ 8s 21ms/step - accuracy: 0.9910 - loss: 0.0305 - val_accuracy: 0.9822 - val_loss: 0.0594
Epoch 3/3
338/338 ━━━━━━━━━━━━━━━━━━━━ 10s 21ms/step - accuracy: 0.9952 - loss: 0.0175 - val_accuracy: 0.9835 - val_loss: 0.0600
169/169 ━━━━━━━━━━━━━━━━━━━━ 1s 8ms/step

Fold 2
Epoch 1/3
338/338 ━━━━━━━━━━━━━━━━━━━━ 10s 23ms/step - accuracy: 0.8667 - loss: 0.3071 - val_accuracy: 0.9822 - val_loss: 0.0592
Epoch 2/3
338/338 ━━━━━━━━━━━━━━━━━━━━ 11s 24ms/step - accuracy: 0.9919 - loss: 0.0271 - val_accuracy: 0.9826 - val_loss: 0.0576
Epoch 3/3
338/338 ━━━━━━━━━━━━━━━━━━━━ 9s 20ms/step - accuracy: 0.9959 - loss: 0.0147 - val_accuracy: 0.9805 - val_loss: 0.0712
169/169 ━━━━━━━━━━━━━━━━━━━━ 1s 8ms/step

Fold 3
Epoch 1/3
338/338 ━━━━━━━━━━━━━━━━━━━━ 10s 24ms/step - accuracy: 0.8731 - 

##### **6. Final Explicit Evaluation on Unseen Test Set**
* Retrain each model configuration explicitly on the full training set, evaluate them on an independent test set, and provide detailed classification metrics along with insightful extreme-error analyses.

In [ ]:
# ---------------------------------------------------------
# Final evaluation for each configuration clearly:
print("\n========== FINAL EVALUATION ON UNSEEN TEST SET ==========\n")

for config_name, (text_data, cluster_data, test_cluster_data) in configurations.items():
    print(f"\n=== Final Classification Report: {config_name} ===")

    # Train on full training data
    if cluster_data is not None:
        model = create_model(cluster_feature_dim=cluster_data.shape[1])
        model.fit(
            [text_data, cluster_data], train_labels,
            epochs=3, batch_size=64, validation_split=0.1
        )
        predictions = (model.predict([X_test_pad, test_cluster_data]) > 0.5).astype("int32")
    else:
        model = create_model()
        model.fit(
            text_data, train_labels,
            epochs=3, batch_size=64, validation_split=0.1
        )
        predictions = (model.predict(X_test_pad) > 0.5).astype("int32")

    # Print final classification report
    print(classification_report(test_labels, predictions, target_names=le.classes_))

    # Extreme error analysis
    error_analysis_df = test_df.copy()
    error_analysis_df['prediction'] = predictions.flatten()
    error_analysis_df['actual'] = test_labels
    extreme_errors = error_analysis_df[error_analysis_df['prediction'] != error_analysis_df['actual']]

    print(f"\nInsightful Analysis of Extreme Errors for {config_name} (Top 10 examples):")
    print(extreme_errors[['email', 'actual', 'prediction']].head(10))


========== FINAL EVALUATION ON UNSEEN TEST SET ==========


=== Final Classification Report: Baseline Bi-LSTM ===
Epoch 1/3
380/380 ━━━━━━━━━━━━━━━━━━━━ 14s 20ms/step - accuracy: 0.8858 - loss: 0.2353 - val_accuracy: 0.9874 - val_loss: 0.0416
Epoch 2/3
380/380 ━━━━━━━━━━━━━━━━━━━━ 7s 19ms/step - accuracy: 0.9919 - loss: 0.0303 - val_accuracy: 0.9889 - val_loss: 0.0346
Epoch 3/3
380/380 ━━━━━━━━━━━━━━━━━━━━ 11s 20ms/step - accuracy: 0.9946 - loss: 0.0181 - val_accuracy: 0.9893 - val_loss: 0.0304
211/211 ━━━━━━━━━━━━━━━━━━━━ 2s 8ms/step
              precision    recall  f1-score   support

         ham       0.99      0.99      0.99      3309
        spam       0.99      0.99      0.99      3434

    accuracy                           0.99      6743
   macro avg       0.99      0.99      0.99      6743
weighted avg       0.99      0.99      0.99      6743


Insightful Analysis of Extreme Errors for Baseline Bi-LSTM (Top 10 examples):
                                                 ema

##### **7. Results**
The experiments evaluated 3 configurations using a Bi-LSTM model with 5-fold cross-validation and an test set.

Cross-validation Results:

* Baseline Bi-LSTM:

    * Accuracy ranged between 98.0–98.7%, with loss consistently low (0.0147–0.0205).

    * Precision, recall, and F1-score were high for both classes (ham: 99% precision, 98% recall, 98% F1; spam: 98% precision, 99% recall, 99% F1).

* Bi-LSTM + K-Means:

    * Similar accuracy (98.1–98.7%) and low loss (0.0136–0.0330) across folds.

    * Kept high metrics with slight variation (ham: 99% precision, 98% recall, 98% F1; spam: 98% precision, 99% recall, 98% F1).

* Bi-LSTM + Hierarchical:

    * Accuracy similarly ranged around 98%, with occasional low loss (0.0157–0.0358).

    * Consistent high performance (ham: 99% precision, 97% recall, 98% F1; spam: 97% precision, 99% recall, 98% F1).

Final Evaluation on Unseen Test Set:

* Baseline Bi-LSTM:

    * Achieved 99% accuracy, very low loss (0.0181), and near-perfect scores (ham and spam: 99% precision, 99% recall, 99% F1).

* Bi-LSTM + K-Means:

    * Accuracy slightly lower (98%), with low loss (0.0153).

    * Precision and recall slightly reduced for ham (99% precision, 97% recall, 98% F1), and for spam (97% precision, 99% recall, 98% F1).

* Bi-LSTM + Hierarchical:

    * Strong performance with 99% accuracy and lowest loss (0.0173), matching the baseline closely.

    * Great precision, recall, and F1-score for both classes (ham: 99% precision, 98% recall, 99% F1; spam: 98% precision, 99% recall, 99% F1).

Extreme Error Analysis:

Analyzing the top misclassifications revealed certain challenging emails consistently misclassified across configurations, indicating subtle linguistic ambiguities rather than model weakness.

Summary:

Overall, all configurations exhibited excellent predictive performance. The hierarchical clustering features performed slightly better than K-Means, but the baseline model already performed as well, emphasizing that clustering features may not add much value compared to just using the baseline model, at least for this dataset.